dataservice/api/common/model.py

```diff
@@ -34,6 +38,111 @@ def kf_id(cls):
     uuid = db.Column(UUID(), unique=True, default=uuid_generator)
 
 
+class IndexdFile:
+    """
+    Field reflection for objects that are stored in indexd
+
+    ### Creation
+
+    When an indexd file is created, an instance of the orm model here is
+    created, and when persisted to the database, a request is sent to Gen3
+    indexd to register the file in the service. Upon successful registry
+    of the file, a response containing a did (digital identifier) will be
+    recieved. The IndexdFile will then be inserted into the database using
+    the baseid as its uuid.
+
+    ### Update
+
+    When a file is updated in indexd, a new version with a new did is created.
+    The document still shares a base_id with the older versions, but a document
+    may not be retrieved with the base_id alone. Because of this, the
+    latest_did is stored on the file.
+
+    ### Deletion
+
+    A file deleted through a DELETE on the dataservice api will immediately
+    delete that file from the dataservice's database, as well as send
+    a corresponding DELETE to the indexd service.
+    Though it should not occur, a file deleted through indexd will remain
+    in the dataservice's database until it a retrieval is attempted.
+    If indexd returns a not found error, the dataservice will automatically
+    remove that file from the database, giving the appearence that the two
+    are in sync from the viewpoint of the dataservice API.
+    """
+    # Store the latest did in the database
+    # files in indexd cannot be looked up by their baseid
+    latest_did = db.Column(UUID(), nullable=False)
+
+    # Fields used by indexd, but not tracked in the database
+    file_name = ''
+    urls = []
+    rev = None
+    hashes = {}
+    # The metadata property is already used by sqlalchemy
+    _metadata = {}
+    size = None
+
+    def merge_indexd(self):
+        """
+        Gets additional fields from indexd
+
+        If the document matching this object's latest_did cannot be found in
+        indexd, remove the object from the database
+
+        :returns: This object, if merge was successful, otherwise None
+        """
+        try:
+            return indexd.get(self)
+        except RecordNotFound as err:
+            self.was_deleted = True
```

Comment:
Are you keeping this flag so that you don't always have to send a request to check if an indexd file was deleted?

Reply:
This is a hack so we don't get stuck in an endless loop in the before_delete event.